### PR TITLE
fix: duplicate logs on toggle

### LINKED ISSUE #26: Miscalculation when log records repeated events and "Show" checkboxes are toggled
Toggling the "Show" checkboxes for log messages cause repeated events (e.g., You made scaffolding (2)) to display inaccurate values.

Original state
image
Deselect "Show Event Messages"
image
Re-select "Show Event Messages"
image

### PR DESCRIPTION
**Pull request by ChandlerFerry**
Fixes #26 

Minor efficiency increase.

Adds a merge interval for back to back messages inside of the `<type>Log`, as we are relying on references in this solution.

### CODE REVIEW

File: src/app/game-state/log.service.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import { Injectable } from '@angular/core';
 import { isEmpty } from 'rxjs';
 
+const LOG_MERGE_INTERVAL_MS = 1000;
```
Comment (ChandlerFerry) [inline src/app/game-state/log.service.ts:4]: Interval to merge messages. Is best tested by `EVENTS`.

image

File: src/app/game-state/log.service.ts [hunk 2]
```diff
@@ -49,20 +50,36 @@ export class LogService {
...
+      // Repeat Found
+      const hasRepeatNumber = /\((\d+)\)$/.exec(log[0].message)
+      let repeatNumber = 2;
+      if(hasRepeatNumber) {
+        repeatNumber = parseInt(hasRepeatNumber[1]) + 1
+      }
+      
+      // Update message reference
+      log[0].message = `${newMessage.message} (${repeatNumber})`;
+    }
```
Comment (ChandlerFerry) [inline src/app/game-state/log.service.ts:76]: Updated & Moved merge logic here.

File: src/app/game-state/log.service.ts [hunk 3]
```diff
@@ -49,20 +50,36 @@ export class LogService {
...
       timestamp: Date.now()
     };
-    log.unshift(newMessage);
+
+    if (log.length == 0 || ((newMessage.timestamp - log[0].timestamp) > LOG_MERGE_INTERVAL_MS) || !log[0].message.includes(newMessage.message)) {
+      // Initialization || Repeat Not Found || Repeat is not within 1 second
+      log.unshift(newMessage);
+      if (this.logTopics.includes(topic)) {
+        this.addToCurrentLog(newMessage);
+      }
```
Comment (ChandlerFerry) [inline src/app/game-state/log.service.ts:65]: Moved `addToCurrentLog` here as we do not want to call the function if it is a merge.

File: src/app/game-state/log.service.ts [hunk 4]
```diff
@@ -75,29 +92,13 @@ export class LogService {
...
-      }
-    } else {
-      this.currentLog.unshift(newMessage);
-      if (this.currentLog.length > 300){
-        // peel off the oldest
-        this.currentLog.splice(this.currentLog.length - 1, 1);
-      }
-      
+  /** Add Message To Current Log */
+  private addToCurrentLog(newMessage: Log): void {
```
Comment (ChandlerFerry) [inline src/app/game-state/log.service.ts:96]: Streamlined as we are relying on references for any merges.

File: src/app/game-state/log.service.ts [hunk 5]
```diff
@@ -41,29 +42,45 @@ export class LogService {
...
+      if (this.logTopics.includes(topic)) {
+        this.addToCurrentLog(newMessage);
+      }
+    } else {
+      // Repeat Found
+      const hasRepeatNumber = /\((\d+)\)$/.exec(log[0].message);
+      let repeatNumber = 2;
+      if(hasRepeatNumber) {
+        repeatNumber = parseInt(hasRepeatNumber[1]) + 1;
+      }
```
Comment (ChandlerFerry) [inline src/app/game-state/log.service.ts:72]: Streamlined solution to determining `repeatNumber`.

**Review by immortalityidle — APPROVED** [review]

**Pull request merged**